File: android/core/src/main/java/ru/maksimbulva/chess/core/engine/move/generator/BaseDeltasMoveGenerator.kt
package ru.maksimbulva.chess.core.engine.move.generator

import ru.maksimbulva.chess.core.engine.board.Cell
import ru.maksimbulva.chess.core.engine.board.Vector2
import ru.maksimbulva.chess.core.engine.move.Move
import ru.maksimbulva.chess.core.engine.move.MoveBuilder
import ru.maksimbulva.chess.core.engine.otherPlayer
import ru.maksimbulva.chess.core.engine.position.Position

internal abstract class BaseDeltasMoveGenerator : IMoveGenerator {

    protected fun generateMoves(
        fromCell: Cell,
        position: Position,
        deltas: Array<Vector2>,
        moves: MutableList<Move>
    ) {
        val board = position.board
        val otherPlayer = position.playerToMove.otherPlayer()
        val moveBuilder = MoveBuilder(fromCell)
        for (delta in deltas) {
            val toCell = fromCell + delta
            if (toCell != null) {
                val moveBuilderForCurrentDelta = moveBuilder.setToCell(toCell)
                if (board.isEmpty(toCell)) {
                    moves.add(moveBuilderForCurrentDelta.build())
                } else {
                    val pieceAtToCell = board.pieceAt(toCell)
                    if (pieceAtToCell?.player == otherPlayer) {
                        moves.add(
                            moveBuilderForCurrentDelta.setAsCapture(pieceAtToCell.piece).build()
                        )
                    }
                }
            }
        }
    }
}

File: ChessApp/ChessEngine/Move/Generator/BaseDeltasMoveGenerator.cs
using ChessEngine.Board;
using System.Collections.Generic;

namespace ChessEngine.Move.Generator
{
    internal abstract class BaseDeltasMoveGenerator : IPieceMoveGenerator
    {
        public void AppendMoves(BoardSquare originSquare, Position.Position position, List<Move> moves)
        {
            var board = position.Board;
            var playerToMove = position.PlayerToMove;
            var moveDeltas = GetMoveDeltas();
            var moveBuilder = new MoveBuilder(GetPiece(), originSquare);

            for (int i = 0; i < moveDeltas.Length; ++i)
            {
                if (!moveDeltas[i].IsCanApplyTo(originSquare))
                {
                    continue;
                }
                var destSquare = moveDeltas[i].GetDestSquare(originSquare);
                if (board.IsEmpty(destSquare))
                {
                    moves.Add(moveBuilder.SetDestSquare(destSquare).Build());
                }
                else
                {
                    var pieceAtDestSquare = board.GetPieceAt(destSquare);
                    if (pieceAtDestSquare.player != playerToMove)
                    {
                        moves.Add(moveBuilder
                            .SetDestSquare(destSquare)
                            .SetCapture(pieceAtDestSquare.piece)
                            .Build());
                    }
                }
            }
        }

        protected abstract MoveDelta[] GetMoveDeltas();

        protected abstract Piece GetPiece();
    }
}
